rust-runtime/aws-smithy-http/src/event_stream/sender.rs

```diff
@@ -3,77 +3,122 @@
 
 /// Input type for Event Streams.
-pub struct EventStreamInput<T> {
-    input_stream: Pin<Box<dyn Stream<Item = Result<T, BoxError>> + Send>>,
+pub struct EventStreamSender<T, E> {
+    input_stream: Pin<Box<dyn Stream<Item = Result<T, E>> + Send>>,
 }
 
-impl<T> fmt::Debug for EventStreamInput<T> {
+impl<T, E> Debug for EventStreamSender<T, E> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "EventStreamInput(Box<dyn Stream>)")
+        write!(f, "EventStreamSender(Box<dyn Stream>)")
     }
 }
 
-impl<T> EventStreamInput<T> {
+impl<T, E: StdError + Send + Sync + 'static> EventStreamSender<T, E> {
     #[doc(hidden)]
-    pub fn into_body_stream<E: StdError + Send + Sync + 'static>(
+    pub fn into_body_stream(
         self,
         marshaller: impl MarshallMessage<Input = T> + Send + Sync + 'static,
+        error_marshaller: impl MarshallMessage<Input = E> + Send + Sync + 'static,
         signer: impl SignMessage + Send + Sync + 'static,
     ) -> MessageStreamAdapter<T, E> {
-        MessageStreamAdapter::new(marshaller, signer, self.input_stream)
+        MessageStreamAdapter::new(marshaller, error_marshaller, signer, self.input_stream)
     }
 }
 
-impl<T, S> From<S> for EventStreamInput<T>
+impl<T, E, S> From<S> for EventStreamSender<T, E>
 where
-    S: Stream<Item = Result<T, BoxError>> + Send + 'static,
+    S: Stream<Item = Result<T, E>> + Send + 'static,
 {
     fn from(stream: S) -> Self {
-        EventStreamInput {
+        EventStreamSender {
             input_stream: Box::pin(stream),
         }
     }
 }
 
+#[derive(Debug)]
+pub struct MessageStreamError {
+    kind: MessageStreamErrorKind,
+    pub(crate) meta: aws_smithy_types::Error,
+}
+
+#[derive(Debug)]
+pub enum MessageStreamErrorKind {
+    Unhandled(Box<dyn std::error::Error + Send + Sync + 'static>),
+}
+
+impl MessageStreamError {
+    /// Creates the `MessageStreamError::Unhandled` variant from any error type.
+    pub fn unhandled(err: impl Into<Box<dyn std::error::Error + Send + Sync + 'static>>) -> Self {
+        Self {
+            meta: Default::default(),
+            kind: MessageStreamErrorKind::Unhandled(err.into()),
+        }
+    }
+
+    /// Creates the `MessageStreamError::Unhandled` variant from a `aws_smithy_types::Error`.
+    pub fn generic(err: aws_smithy_types::Error) -> Self {
+        Self {
+            meta: err.clone(),
+            kind: MessageStreamErrorKind::Unhandled(err.into()),
+        }
+    }
+
+    /// Returns error metadata, which includes the error code, message,
+    /// request ID, and potentially additional information.
+    pub fn meta(&self) -> &aws_smithy_types::Error {
+        &self.meta
+    }
+}
+
+impl StdError for MessageStreamError {}
+impl fmt::Display for MessageStreamError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.kind {
+            MessageStreamErrorKind::Unhandled(inner) => std::fmt::Debug::fmt(inner, f),
```

Comment:
Shouldn't this be formatting it with `Display` instead?

Reply:
Yes thanks